pkg/cloud/ibmcloud/clients/machines.go

```diff
@@ -176,32 +177,30 @@ func (gs *GuestService) DeleteGuest(Id int) error {
 	return err
 }
 
-func (gs *GuestService) listGuest() ([]datatypes.Virtual_Guest, error) {
+func (gs *GuestService) GetGuest(name string) (*datatypes.Virtual_Guest, error) {
 	s := services.GetAccountService(gs.sess)
 
-	guests, err := s.GetVirtualGuests()
+	hostFilter := filter.Build(
+		filter.Path("virtualGuests.hostname").Eq(name),
+	)
+
+	guests, err := s.Filter(hostFilter).GetVirtualGuests()
 	if err != nil {
-		klog.Errorf("Error listing virtual guest: %v", err)
+		klog.Errorf("Error getting virtual guests: %v", err)
 		return nil, err
 	}
-	return guests, nil
-}
 
-// FIXME: use API layer query instead of query all then compare here
-func (gs *GuestService) GetGuest(name string) (*datatypes.Virtual_Guest, error) {
-	guests, err := gs.listGuest()
-	if err != nil {
-		return nil, err
+	if len(guests) == 0 {
+		return nil, nil
 	}
 
-	for _, guest := range guests {
-		// FIXME: how to unique identify one guest
-		if *guest.Hostname == name {
-			klog.Infof("Found guest with ID %d for %q", *guest.Id, name)
-			return &guest, nil
-		}
+	if len(guests) > 1 {
+		// I noticed that IBM Cloud can use same name for 2 machines.
+		// It is bad for our case. Print a message to make it to be noticed.
+		klog.Errorf("Getting more than one virtual guest. The first is used.")
```

Comment:
Print out the guest name here for better debug.

Reply:
I use guest ID because names are the same when issue happens.